README.md
# deno-statsd-client

[![Build Status](https://travis-ci.com/raydog/deno-statsd-client.svg?branch=main)](https://travis-ci.com/raydog/deno-statsd-client)

A simple StatsD client for Deno.

Supports the official stat metrics (counts, timings, gauges, sets), a few
datadog extensions (histogram, distribution, events, service checks) as well as
both TCP, UDP, and unix domain socket connections.

**Deno command-line args**:

You'll need certain flags to use this library, either to enable certain
permsissions, or to enable unstable APIs:

| If connecting via:  | You'll need:                                |
| ------------------- | ------------------------------------------- |
| UDP Datagrams       | `--unstable` `--allow-net`                  |
| TCP Packets         | `--unstable` `--allow-net`                  |
| Unix Domain Sockets | `--unstable` `--allow-read` `--allow-write` |

```ts
import { StatsDClient } from "https://deno.land/x/statsd@0.5.0/mod.ts";

const client = new StatsDClient({
  server: {
    proto: "udp",
    host: "localhost",
    port: 8125,
  },
});

// Count an event:
client.count("http.routes.get_resource");

// Track how long something took:
client.timing("http.routes.get_resource.200", responseTime);

// Track some interesting global measurement:
client.gauge(`server.${Deno.hostname()}.disk_left`, await getDiskRemaining());

// Adjust some other global measurement:
client.adjustGauge("assets.total_bytes_stored", new_upload.byteLength);

// Keep track of how many distinct things we've seen:
client.unique("users.unique", user.id);
```

## Constructor options:

- `server` (Object?) Object that describes how to connect to the server.

  If connecting via UDP datagrams:

  - `proto`: `"udp"`
  - `host` (string?) (Default: "localhost")
  - `port` (number?) (Default: 8125)
  - `mtu` (number?) (Default: 1500)

    The Maximum Transmission Unit for the network connection.

    We use this number when figuring out the maximum amount of data that we can
    send in a single network packet. A smaller number means more packets will
    have to be sent, but if we set this value _TOO_ high, it might mean that
    packets won't arrive.

    1500 bytes is usually safe enough for most server networks. Fancy networks
    that have Jumbo Frames enabled might be able to bump this value higher, like
    to 8932 bytes, but worse networks (like if these packets are routed through
    the wider internet) might need to reduce the MTU to 512 or less. It all
    depends on the routers that these packets get routed through, and how they
    were configured.

  If connecting via a TCP socket:

  - `proto`: `"tcp"`
  - `host` (string?) (Default: "localhost")
  - `port` (number?) (Default: 8125)
  - `maxQueue` (number?) (Default: 100)

    Sent metrics are queued up for a short bit (see: maxDelayMs) before sending
    to increase the number of metrics in each TCP frame. However, if the backlog
    exceeds this number of metrics, we'll send the items sooner.

  If connecting via a Unix domain socket:

  - `proto`: `"unix"`
  - `path` (string)
  - `maxQueue` (number?) (Default: 100)

    Sent metrics are queued up for a short bit (see: maxDelayMs) before sending,
    to decrease the number of filesystem writes. However, if the backlog exceeds
    this number of items, we'll send the items sooner.

  If you don't _actually_ want to connect, but instead just log every metric:

  - `proto`: `"logger"`

- `sampleRate` (number?) (Default: 1.0)

  The sampling rate we'll use for metrics. This should be a value between 0 and
  1, inclusive.

  For example, if this value is set to 0.1, then 1 out of 10 calls to .count()
  will actually result in a counter being sent to the server. HOWEVER, the
  server will then increase the counter by 10x the amount it normally would have
  been increased. This will result in less data being sent over the wire, but
  with mostly the same ending values. (Albeit with a bit more error.)

- `safeSampleRate` (boolean?) (Default: true)

  StatsD occasionally has some erratic behaviors when dealing with sampleRates.
  For example, relative gauges don't have any sampleRate corrections on the
  server-side, and so would result in the wrong number of adjustments being made
  to the data. Same with sets: the wrong number of unique values will be
  reported if sampleRates are used.

  This setting, when true, will cause us to ignore the sampleRate in metrics
  that wouldn't handle it well. (Relative gauges, and Sets.)

- `maxDelayMs` (number?) (Default: 1000)

  When we get a metric to send, we'll wait (at most) this number of milliseconds
  before actually sending it. This gives us some time for other metrics to be
  queued up, so we can send them all at once, in the same packet. We may decide
  to send the packet sooner (like if it gets too big for the MTU) but in
  general, this is the maximum amount of time that your metric will be delayed.

- `globalTags`: (Object?) (Default: `{}`)

  Tags are key-value pairs that are appended to each metric. Values need to be
  strings. If a value is the empty string, that tag will be skipped.

- `logger`: (Logger?) (Default: no logging)

  This library will not log anything by default. However, if you'd get a little
  more feedback about how the StatsD connection is going, you can get that by
  providing a logger object here. The typescript types of this parameter is
  pretty general, but should be compatible with most versions of the std
  library's logger:

  ```ts
  import * as log from "https://deno.land/std@0.95.0/log/mod.ts";

  await log.setup({ ... })

  const c = new StatsDClient({
    logger: log.getLogger("statsd"),
  });
  ```

## Client Methods

- `count(key: string, num, opts?: MetricOpts)`

  Send a "count" metric. This is used to track the number of things.

  ```ts
  // Count the number of times a route was used:
  client.count("routes.get_api_resource");

  // Count the number of items purchased:
  client.count("store.widgets.purchased", order.items.length);
  ```

- `timing(key: string, ms: number, opts?: MetricOpts)`

  Sends a "timing" metric, typically measured in milliseconds. The remote StatsD
  server will usually calculate other metrics based on these values. These
  metrics can include things like: min, max, average, mean90, etc...

  While this metric type was originally intended only for timing measurements,
  it can really be used for any value where things like min, max, mean90, etc...
  would be useful.

  ```ts
  // Keep track of route response times:
  client.timing("routes.get_api_resource.ok", Date.now() - start);
  ```

- `gauge(key: string, value: number, opts?: MetricOpts)`

  Sends a "gauge" metric. Gauges are point-in-time measurements, that are true,
  at this time. The StatsD server will keep gauges in memory, and will keep
  using them if a new value wasn't sent. Use this to track values that are
  absolutely true, at this point-in-time. This could include things like "number
  of items in a db table", or "bytes remaining in the main disk partition."
  Things like that.

  ```ts
  // Keep track of server disk usage:
  client.gauge(
    `servers.${await Deno.hostname()}.diskPercent`,
    await getDiskPercent(),
  );

  // Keep track of items in a db table:
  client.gauge("database.main.users", await userTable.count());
  ```

- `adjustGauge(key: string, delta: number, opts?: MetricOpts)` Sends a relative
  "gauge" metric. A _relative_ gauge metric may reference a gauge value (an
  absolute measurement) but we aren't sending that exact measurement right now.
  We're just sending an adjustment value.

  ```ts
  // Adjust the total asset size after an upload:
  client.adjustGauge("assets.size.overall", asset.byteLength);

  // Adjust the total number of users after a cancellation:
  client.adjustGauge("users.count", -1);
  ```

- `unique(key: string, value: number | string, opts?: MetricOpts)`

  Sends a "set" metric. A set metric tracks the number of distinct values seen
  in StatsD over time. Use this to track things like the number of disctint
  users.

  ```ts
  // Track distinct authenticated users
  client.unique("users.distinct", user.id);
  ```

- `histogram(key: string, value: number, opts?: MetricOpts)`

  (Datadog-only)

  Sends a "histogram" metric. A histogram is pretty much the same thing as a
  timer, but created by datadog, and not compatible with StatsD.

  I don't get it either.

- `distribution(key: string, value: number, opts?: MetricOpts)`

  (Datadog-only)

  Sends a "distribution" metric. A distribution seems to be functionally
  equivalent to a timing metric, but has its own page in datadog that has been
  deprecated.

- `event(event: EventOpts)`

  (Datadog-only)

  Sends an "event" to datadog.

  ```ts
  // Send an exception to datadog:
  client.event({
    title: "Unexpected Error",
    text: err.stack || "",
    aggregate: "Errors",
    source: "database",
    type: "error",
  });
  ```

  This method only takes a single argument with these properties:

  - `title` (string)

    Title of the event.

  - `text` (string)

    Event text.

  - `time` (Date?) (Default: `new Date()`)

    Timestamp for the event.

    If omitted, we'll use the current time, so only send this if you're
    submitting an event retroactively.

  - `host` (string | false) (Default: `Deno.hostname()`)

    The hostname for the event.

    If omitted, we'll use `Deno.hostname()` if the current process has
    environment permissions. Else the param will not be sent with the event. If
    you wish to omit hostname from the event, use the value `false`.

  - `aggregate` (string?)

    An aggregation key for the event. Similar events will be grouped in datadog
    by this key.

  - `priority`: ("normal" | "low") (Default: "normal")

    A priority level for this event.

  - `source`: (string?)

    A source name for this event.

  - `type`: ("error" | "warning" | "info" | "success")

    An alert type for this event.

  - `tags`: (Object?)

    Tags that will be attached to this event.

- `serviceCheck(check: ServiceCheckOpts)`

  (Datadog-only)

  Sends a "service check" to datadog.

  ```ts
  // Send a critical redis outage to datadog.
  client.serviceCheck({
    name: "Redis connection",
    status: "critical",
    tags: { env: "production" },
    message: "Redis connection timed out after 10s",
  });
  ```

  This method only takes a single argument with these properties:

  - `name` (string)

    Name of the service check

  - `status`: ("ok" | "warning" | "critical" | "unknown")

    Service check status.

  - `time` (Date?) (Default: `new Date()`)

    Timestamp for the service check.

    Default is "now", so only supply this if submitting a service check
    retroactively.

  - `host` (string | false) (Default: `Deno.hostname()`)

    The hostname for this service check.

    If omitted, we'll use `Deno.hostname()` if the current process has
    environment permissions. Else the param will not be sent with the event. If
    you wish to omit hostname from the event, use the value `false`.

  - `tags`: (Object?)

    Tags for this service check.

  - `message`: (string?)

    A message describing the current status.

- `async close(): Promise<void>`

  Will flush all pending metrics, and close all open sockets.

  Any attempts to use this client after close() should error.

### MetricOpts

Some params can be overridden for each metric with the MetricOpts object.

- `sampleRate` - To use a custom sampleRate for this metric.
- `tags` - To add some extra tags to this metric. (They are merged with the
  global tags)


HISTORY.md
# Version history

## v0.5.0 (2021-05-06)

- Allow custom loggers to be injected. Stop relying on std.

## v0.4.1 (2021-04-13)

- Bump std to 0.93.0.
- Fix a minor type issue that cropped up with Deno 1.9.0.

## v0.4.0 (2021-04-11)

- Can now select a "dialect" of the StatsD service. (Currently, "statsd" or
  "datadog".)
- Added datadog-specific metrics and validation.
- Added datadog events.
- Added datadog service checks.
- Added a new "logger" server config for debugging.

## v0.3.0 (2021-04-07)

- Add UDP (Unix Domain Socket) support.

## v0.2.0 (2021-04-06)

- Add support for TCP connections
- Can now enabling logging by configuring the `statsd` logger in the std logging
  library.

## v0.1.1 (2021-04-04)

- Minor type and doc fixes

## v0.1.0 (2021-04-04)

- Initial version
- Supports UDP connections
- Supports all official StatsD metrics (count, gague, timing, set)


src/StatsDClient.ts
import { Client } from "./types/Client.ts";
import { UDPClient } from "./network/UDPClient.ts";
import { SocketClient } from "./network/SocketClient.ts";
import {
  LibConfig,
  TCPConfig,
  UDPConfig,
  UnixConfig,
} from "./types/LibConfig.ts";
import { MetricOpts } from "./types/MetricOpts.ts";
import * as formats from "./utils/formats.ts";
import { StatsDError } from "./StatsDError.ts";
import { Dialect } from "./types/Dialect.ts";
import { StatsDDialect } from "./dialects/StatsDDialect.ts";
import { DatadogDialect } from "./dialects/DatadogDialect.ts";
import { EventOpts, InternalEventOpts } from "./types/EventOpts.ts";
import { LoggerClient } from "./network/LoggerClient.ts";
import {
  InternalServiceCheckOpts,
  ServiceCheckOpts,
} from "./types/ServiceCheckOpts.ts";
import { Tags } from "./types/Tags.ts";
import { Logger } from "./types/Logger.ts";

/**
 * StatsD client. Use this to send data to a StatsD-speaking backend.
 */
export class StatsDClient {
  #client: Client | null;

  #globalOpts: Required<MetricOpts>;
  #safeSampleRate: boolean;

  #dialect: Dialect;
  #cachedHostname: string | undefined;
  #logger: Logger;

  // Returns the client. Used to implement a shutdown state:
  private getClient(): Client {
    if (!this.#client) {
      throw new StatsDError("Client is closed");
    }
    return this.#client;
  }

  // Fetch the current hostname. Cache the result.
  // If permissions haven't been granted, we'll cache "".
  private getHostname(): string {
    if (this.#cachedHostname !== undefined) return this.#cachedHostname;

    try {
      this.#cachedHostname = Deno.hostname();
      return this.#cachedHostname;
    } catch (ex) {
      if (ex instanceof Deno.errors.PermissionDenied) {
        this.#cachedHostname = "";
        return this.#cachedHostname;
      }
      throw ex;
    }
  }

  /**
   * Build a new client. This client will have its own connection to the remote StatsD server.
   * 
   * If no config object is passed in, we'll 100% fall back on default settings. (localhost:8125, UDP, ...)
   * 
   * @param conf Settings.
   */
  constructor(conf?: LibConfig) {
    this.#globalOpts = {
      sampleRate: conf?.sampleRate ?? 1.0,
      tags: conf?.globalTags ?? {},
    };
    this.#safeSampleRate = conf?.safeSampleRate ?? true;
    this.#dialect = _getDialect(conf);
    this.#logger = conf?.logger || {
      debug() {},
      info() {},
      warning() {},
      error() {},
    };
    const maxDelay = conf?.maxDelayMs ?? 1000;
    this.#client = _connect(conf?.server, maxDelay, this.#logger);
  }

  /**
   * Send a "count" metric. This is used to track the number of things.
   * 
   * @example
   *   // Count the number of times a route was used:
   *  client.count("routes.get_api_resource");
   * 
   * @example
   *   // Count the number of items purchased:
   *   client.count("store.widgets.purchased", order.items.length);
   * 
   * @param key  Metric key
   * @param num  Number of things (Default is 1)
   * @param opts Extra options
   */
  count(key: string, num = 1, opts?: MetricOpts) {
    const client = this.getClient();
    const sample = _getSampling(
      this.#globalOpts,
      opts,
      true,
      this.#safeSampleRate,
    );
    const tags = _getTags(this.#globalOpts, opts);
    if (!_doSampling(sample)) return;
    const data = formats.buildCountBody(
      this.#dialect,
      key,
      num,
      sample,
      tags,
    );
    client.queueData(data);
  }

  /**
   * Sends a "timing" metric, typically measured in milliseconds. The remote StatsD server will usually calculate other
   * metrics based on these values. These metrics can include things like: min, max, average, mean90, etc...
   * 
   * While this metric type was originally intended only for timing measurements, it can really be used for any value
   * where things like min, max, mean90, etc... would be useful.
   * 
   * @example
   *   // Keep track of route response times:
   *   client.timing("routes.get_api_resource.ok", Date.now() - start);
   * 
   * @param key Metric key
   * @param ms  Measurement
   * @param opts Extra options
   */
  timing(key: string, ms: number, opts?: MetricOpts) {
    const client = this.getClient();
    const sample = _getSampling(
      this.#globalOpts,
      opts,
      true,
      this.#safeSampleRate,
    );
    const tags = _getTags(this.#globalOpts, opts);
    if (!_doSampling(sample)) return;
    const data = formats.buildTimingBody(
      this.#dialect,
      key,
      ms,
      sample,
      tags,
    );
    client.queueData(data);
  }

  /**
   * Sends a "gauge" metric. Gauges are point-in-time measurements, that are true, at this time. The StatsD server will
   * keep gauges in memory, and will keep using them if a new value wasn't sent. Use this to track values that are
   * absolutely true, at this point-in-time. This could include things like "number of items in a db table", or "bytes
   * remaining in the main disk partition." Things like that.
   * 
   * @example
   *   // Keep track of server disk usage:
   *   client.gauge(`servers.${await Deno.hostname()}.diskPercent`, await getDiskPercent());
   * 
   * @example
   *   // Keep track of items in a db table:
   *   client.gauge("database.main.users", await userTable.count());
   * 
   * @param key   Metric key
   * @param value Measurement
   * @param opts  Extra options
   */
  gauge(key: string, value: number, opts?: MetricOpts) {
    const client = this.getClient();
    const sample = _getSampling(
      this.#globalOpts,
      opts,
      true,
      this.#safeSampleRate,
    );
    const tags = _getTags(this.#globalOpts, opts);
    if (!_doSampling(sample)) return;
    const data = formats.buildAbsGaugeBody(
      this.#dialect,
      key,
      value,
      sample,
      tags,
    );
    client.queueData(data);
  }

  /**
   * Sends a relative "gauge" metric. A _relative_ gauge metric may reference a gauge value (an absolute measurement)
   * but we aren't sending that exact measurement right now. We're just sending an adjustment value.
   * 
   * @example
   *   // Adjust the total asset size after an upload:
   *   client.adjustGauge("assets.size.overall", asset.byteLength);
   * 
   * @example
   *   // Adjust the total number of users after a cancellation:
   *   client.adjustGauge("users.count", -1);
   * 
   * @param key   Metric key
   * @param delta Adjustment value
   * @param opts  Extra options
   */
  adjustGauge(key: string, delta: number, opts?: MetricOpts) {
    const client = this.getClient();
    const sample = _getSampling(
      this.#globalOpts,
      opts,
      false,
      this.#safeSampleRate,
    );
    const tags = _getTags(this.#globalOpts, opts);
    if (!_doSampling(sample)) return;
    const data = formats.buildRelGaugeBody(
      this.#dialect,
      key,
      delta,
      sample,
      tags,
    );
    client.queueData(data);
  }

  /**
   * Sends a "set" metric. A set metric tracks the number of distinct values seen in StatsD over time. Use this to track
   * things like the number of disctint users.
   * 
   * @example
   *   // Track distinct authenticated users
   *   client.unique("users.distinct", user.id);
   * 
   * @param key   Metric key
   * @param value Identifying value
   * @param opts  Extra options
   */
  unique(key: string, value: number | string, opts?: MetricOpts) {
    const client = this.getClient();
    const sample = _getSampling(
      this.#globalOpts,
      opts,
      false,
      this.#safeSampleRate,
    );
    const tags = _getTags(this.#globalOpts, opts);
    if (!_doSampling(sample)) return;
    const data = formats.buildSetBody(
      this.#dialect,
      key,
      value,
      sample,
      tags,
    );
    client.queueData(data);
  }

  /**
   * Sends a "histogram" metric. A histogram is pretty much the same thing as a timer, but created by datadog, and not
   * compatible with StatsD.
   * 
   * I don't get it either.
   * 
   * This extension to the StatsD protocol is only available when the Dialect is "datadog". Normal StatsD setups should
   * just use the timer metric.
   * 
   * @see Docs https://docs.datadoghq.com/developers/metrics/dogstatsd_metrics_submission/#histogram
   * @deprecated
   * 
   * @param key   Metric key
   * @param value Timing value
   * @param opts  Extra options
   */
  histogram(key: string, value: number, opts?: MetricOpts) {
    const client = this.getClient();
    const sample = _getSampling(
      this.#globalOpts,
      opts,
      true,
      this.#safeSampleRate,
    );
    const tags = _getTags(this.#globalOpts, opts);
    if (!_doSampling(sample)) return;
    const data = formats.buildHistogramBody(
      this.#dialect,
      key,
      value,
      sample,
      tags,
    );
    client.queueData(data);
  }

  /**
   * Sends a "distribution" metric. A distribution seems to be functionally equivalent to a timing metric, but has its
   * own page in datadog that has been deprecated.
   * 
   * This extension to the StatsD protocol is only available when the Dialect is "datadog". Normal StatsD setups should
   * just use the timer metric.
   * 
   * @see Docs https://docs.datadoghq.com/developers/metrics/dogstatsd_metrics_submission/#distribution
   * @deprecated
   * 
   * @param key   Metric key
   * @param value Timing value
   * @param opts  Extra options
   */
  distribution(key: string, value: number, opts?: MetricOpts) {
    const client = this.getClient();
    const sample = _getSampling(
      this.#globalOpts,
      opts,
      true,
      this.#safeSampleRate,
    );
    const tags = _getTags(this.#globalOpts, opts);
    if (!_doSampling(sample)) return;
    const data = formats.buildDistributionBody(
      this.#dialect,
      key,
      value,
      sample,
      tags,
    );
    client.queueData(data);
  }

  /**
   * Sends an "event" to datadog.
   * 
   * This method will only work if the dialect is "datadog".
   * 
   * @example
   *   // Send an exception to datadog:
   *   client.event({
   *     title: "Unexpected Error",
   *     text: err.stack || "",
   *     aggregate: "Errors",
   *     source: "database",
   *     type: "error",
   *   });
   * 
   * @param event Event options
   */
  event(event: EventOpts) {
    const client = this.getClient();
    const host = (typeof event.host === "string")
      ? event.host
      : (event.host === false)
      ? ""
      : this.getHostname();
    const ev: InternalEventOpts = {
      title: event.title,
      text: event.text,
      time: event.time ?? new Date(),
      host,
      aggregate: event.aggregate,
      priority: event.priority ?? "normal",
      source: event.source,
      type: event.type ?? "info",
      tags: _getTags(this.#globalOpts, event),
    };
    const data = formats.buildEventBody(this.#dialect, ev);
    client.queueData(data);
  }

  /**
   * Sends a "service check" to datadog.
   * 
   * This method will only work if the dialect is "datadog".
   * 
   * @example
   *   // Send a critical redis outage to datadog.
   *   client.serviceCheck({
   *     name: "Redis connection",
   *     status: "critical",
   *     tags: { env: "production" },
   *     message: "Redis connection timed out after 10s"
   *   });
   * 
   * @param check Service check options
   */
  serviceCheck(check: ServiceCheckOpts) {
    const client = this.getClient();
    const host = (typeof check.host === "string")
      ? check.host
      : (check.host === false)
      ? ""
      : this.getHostname();
    const opts: InternalServiceCheckOpts = {
      name: check.name,
      status: check.status,
      time: check.time ?? new Date(),
      host,
      tags: _getTags(this.#globalOpts, check),
      message: check.message ?? "",
    };
    const data = formats.buildServiceCheckBody(this.#dialect, opts);
    client.queueData(data);
  }

  /**
   * Will flush all pending metrics, and close all open sockets.
   * 
   * Any attempts to use this client after close() should error.
   */
  async close(): Promise<void> {
    const client = this.getClient();
    this.#client = null;
    await client.close();
  }
}

// Get the effective sampling rate for this metric:
function _getSampling(
  globalOpts: Required<MetricOpts>,
  metricOpts: MetricOpts | undefined,
  isSafe: boolean,
  forceSafe: boolean,
) {
  // If this is a non-safe metric in safe mode, always send:
  if (!isSafe && forceSafe) return 1.0;
  return metricOpts?.sampleRate ?? globalOpts.sampleRate;
}

// Get the full tag set for this metric:
function _getTags(
  globalOpts: Required<MetricOpts>,
  metricOpts: MetricOpts | undefined,
): Tags {
  return {
    ...globalOpts.tags,
    ...metricOpts?.tags,
  };
}

// Return true if we should send this metric, based on the sampling rates, and the metric in question:
function _doSampling(rate: number): boolean {
  return (rate >= 1 || Math.random() < rate);
}

function _connect(
  info: LibConfig["server"],
  maxDelay: number,
  logger: Logger,
): Client {
  const proto = info?.proto || null;
  switch (proto) {
    case null:
    case "udp": {
      const udp = info as UDPConfig;
      const host = udp?.host ?? "localhost";
      const port = udp?.port ?? 8125;
      const mtu = udp?.mtu ?? 1500;
      return new UDPClient({ host, port, mtu, maxDelay, logger });
    }

    case "tcp": {
      const tcp = info as TCPConfig;
      const host = tcp.host ?? "localhost";
      const port = tcp.port ?? 8125;
      const maxQueue = tcp.maxQueue ?? 100;
      return new SocketClient({
        mode: "tcp",
        host,
        port,
        maxQueue,
        maxDelay,
        logger,
      });
    }

    case "unix": {
      const tcp = info as UnixConfig;
      const path = tcp.path;
      const maxQueue = tcp.maxQueue ?? 100;
      return new SocketClient({
        mode: "unix",
        path,
        maxQueue,
        maxDelay,
        logger,
      });
    }

    case "logger":
      return new LoggerClient({ logger });
  }
}

function _getDialect(config: LibConfig | undefined) {
  switch (config?.dialect) {
    case undefined:
    case "statsd": {
      const proto = config?.server?.proto ?? "udp";
      return new StatsDDialect(proto === "udp");
    }

    case "datadog":
      return new DatadogDialect();
  }
}
